(fix): HTTP error handling in VC API

protocolwhisper opened the pull request:
## Issue Addressed

Closes #4597 

## Proposed Changes

Updated Handler Return Types: Handlers now return `Response` directly
Refactored blocking_signed_json_task function

## Additional Info

No

protocolwhisper committed:
(fix): HTTP error handling in VC API

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=4606) <br/>All committers have signed the CLA.

protocolwhisper commented:
@michaelsproul, since [this PR](https://github.com/sigp/lighthouse/pull/4595) got merged, shall I update to use the function `convert_rejection<T>` as a crate in this scope?

michaelsproul commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1226,38 +1247,38 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
+            let response_body = match serde_json::to_vec(&func_output) {
+                Ok(body) => body,
+                Err(_) => {
+                    return WarpResponse::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from(vec![]))
+                        .expect("can produce simple response from static values")
```
By returning this error early we avoid signing it, which is a regression compared to the previous behaviour. Later if we convert an error to a response with `convert_rejection` we also don't sign that, when we should.

Let's refactor this so that we first produce a `Result<Response, Rejection>` from `warp_utils::task::blocking_task`. Then we run `convert_rejection` to get a `Response`, and then finally we sign the body of that response and append the signature to the headers using `headers_mut`. I.e. re-use this block of code from the old implementation, but only after calling `convert_rejection`.

https://github.com/sigp/lighthouse/blob/c7e978b727ec7dbc62951af5e92e6d5413e4c74e/validator_client/src/http_api/mod.rs#L1254-L1259

protocolwhisper replied:
Thanks for the clarification :green_heart:.It makes more sense now., I've just created a PR following it.

michaelsproul commented:
> @michaelsproul, since https://github.com/sigp/lighthouse/pull/4595 got merged, shall I update to use the function convert_rejection<T> as a crate in this scope?

I think it's OK to duplicate the definition for now. Or we could move `convert_rejection` into `warp_utils`.

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
(rebase): HTTP error handling in VC API

jimmygchen commented:
@protocolwhisper looks like there are a few CI failures, would you mind taking a look?

protocolwhisper committed:
(rebase) : Not working

protocolwhisper commented:
@michaelsproul @jimmygchen I've followed your directions, but I'm still getting errors. My function now returns a resp. I've made changes to the blocking_task and another function, but I'm unsure. I'd appreciate some guidance on this.

protocolwhisper committed:
Revert "(rebase) : Not working"

This reverts commit 24fa06bf78b7bb1a710ddb8a6fc083d38d96a17d.

protocolwhisper commented:
@michaelsproul @jimmygchen Hello! I'm currently facing an issue. I've made changes to the function as follows:
However, after making these changes, I'm encountering numerous errors, and I'm not sure where to start in order to resolve them. I'd truly appreciate some guidance on this matter. 
```rust
pub async fn blocking_signed_json_task<S, F, T>(signer: S, func: F) -> Response<Body>
where
    S: Fn(&[u8]) -> String,
    F: FnOnce() -> Result<T, Rejection> + Send + 'static,
    T: Reply + Send + 'static,
{
    // Get the Result from the blocking task
    let result = warp_utils::task::blocking_task(func).await;

    // Convert the result using your convert_rejection function
    let mut response = convert_rejection(result).await;

    // Process the response for the signature
    let body: &Vec<u8> = response.body(); // NOTE: Ensure `body` method exists and returns a Vec<u8>
    let signature = signer(body);
    let header_value = HeaderValue::from_str(&signature).expect("hash can be encoded as header");
    response.headers_mut().append("Signature", header_value);

    response
}

protocolwhisper committed:
(rebase) : Http vc api

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1223,6 +1225,37 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
+///
+/// This function should *always* be used to convert rejections into responses. This prevents warp
+/// from trying to backtrack in strange ways. See: https://github.com/sigp/lighthouse/issues/3404
+pub async fn convert_with_header<T: Reply>(res: Result<T, warp::Rejection>) -> resp {
```
It seems like this function isn't used anywhere and can be removed? (There are some compilation errors as well)

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1235,29 +1268,10 @@ where
-            response
-        })
+    let result = warp_utils::task::blocking_task(func).await;
+    let conversion = convert_rejection(result).await; // It handles the rejection
```
I'm not following the logic here, `convert_rejection` isn't available in the scope, perhaps you're trying to call the `convert_with_header` you defined above?

protocolwhisper replied:
Right, I’ve already modified it but haven’t uploaded the last commit. I’m using ‘convert_with_header’ instead of ‘convert_rejection

protocolwhisper replied:
I've made significant progress. I'll check tomorrow to see if only the `.then` is causing the failure, and then I'll modify the handlers that call `blocking_signed_task` :)

jimmygchen commented:
@protocolwhisper I think it's best to go through the compiler errors - the explanations are usually pretty clear, there are a few easy ones to fix, e.g. method not found, variable not found, I'd suggest to fix those first if you're not sure where to start. I've pointed out 2 of these in the above comments, let us know if there's any specific error that you're struggling to resolve.

protocolwhisper commented:
@jimmygchen The problem arises when running ‘cargo test’ inside the module. I get a ton of errors. I believe they exist because we changed the result of the function to a Response. I wrote a file detailing these errors since they are extensive. Can I upload that file?

protocolwhisper committed:
(rebase): block_signed_task ,  blocking_task

protocolwhisper committed:
(Fix) : Rejection handling in VC API

protocolwhisper commented:
@michaelsproul @jimmygchen I've updated the latest PR with the necessary changes and it now passes the tests for the module. Let me know your feedback :) .

protocolwhisper commented:
Hello 👋🏽, can I get your comments on this PR? I’d like to make changes since the function’s signature has been modified. @michaelsproul @jimmygchen

jimmygchen commented:
Hi @protocolwhisper, thanks for the update!

It looks like @michaelsproul has triggered a CI run, and there are some CI failures *possibly* related to the changes, would you mind taking a look please?
- failing tests: https://github.com/sigp/lighthouse/actions/runs/5934547690/job/16091676680?pr=4606
- clippy issues: https://github.com/sigp/lighthouse/actions/runs/5934547690/job/16091678636?pr=4606 (you can run `make lint` locally to check as well)

protocolwhisper commented:
> Hi @protocolwhisper, thanks for the update!
> 
> It looks like @michaelsproul has triggered a CI run, and there are some CI failures _possibly_ related to the changes, would you mind taking a look please?
> 
> * failing tests: https://github.com/sigp/lighthouse/actions/runs/5934547690/job/16091676680?pr=4606
> * clippy issues: https://github.com/sigp/lighthouse/actions/runs/5934547690/job/16091678636?pr=4606 (you can run `make lint` locally to check as well)

Gotcha, I corrected the lint but since we changed the signature from `pub async fn blocking_signed_json_task<S, F, T>(
    signer: S,
    func: F,
) -> Result<impl warp::Reply, warp::Rejection>`  to `pub async fn blocking_signed_json_task<S, F, T>(signer: S, func: F) -> resp` is this why the test are failing? So I should focus on those changes or the tests? I'm a little lost. When trying to reproduce I'm getting this : 
![image](https://github.com/sigp/lighthouse/assets/57886661/452cf585-55a8-4e6e-a257-546884655d79)

protocolito committed:
(rebase) : Fix lint

protocolito committed:
Revert "(rebase) : Fix lint"

This reverts commit f05c3fbb76a970965af4cb198d7f7af03c6d81b4.

protocolwhisper committed:
(rebase): Fix Lint

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

jimmygchen commented:
Hi @protocolwhisper 
It looks like you don't have [anvil](https://github.com/foundry-rs/foundry/tree/master/crates/anvil) installed, which is required for some tests.

There are a few other requirements for running the tests locally, please see the [Development Envirnonment setup page](https://lighthouse-book.sigmaprime.io/setup.html).

The tests ensure that new changes don't break existing functionalities - if the tests are failing due to changes in method signature it uses directly, they might need to be updated to make the tests pass; or if the changes break existing functionalities, then they need to be fixed.

Thanks - let me know if you run into other issues.

protocolwhisper commented:
> Hi @protocolwhisper It looks like you don't have [anvil](https://github.com/foundry-rs/foundry/tree/master/crates/anvil) installed, which is required for some tests.
> 
> There are a few other requirements for running the tests locally, please see the [Development Envirnonment setup page](https://lighthouse-book.sigmaprime.io/setup.html).
> 
> The tests ensure that new changes don't break existing functionalities - if the tests are failing due to changes in method signature it uses directly, they might need to be updated to make the tests pass; or if the changes break existing functionalities, then they need to be fixed.
> 
> Thanks - let me know if you run into other issues.

I already installed the Dev Dependencies but know I'm running into this issue, even if I did it in the lighthouse unstable fresh : 
![image](https://github.com/sigp/lighthouse/assets/57886661/17db8b93-d0c9-4a36-b088-39b944bb3d8d)

jimmygchen commented:
> I already installed the Dev Dependencies but know I'm running into this issue, even if I did it in the lighthouse unstable fresh : ![image](https://user-images.githubusercontent.com/57886661/264223579-17db8b93-d0c9-4a36-b088-39b944bb3d8d.png)

What OS and rust version (`rustc --version`) are you running the tests with?

protocolwhisper commented:
> > I already installed the Dev Dependencies but know I'm running into this issue, even if I did it in the lighthouse unstable fresh : ![image](https://user-images.githubusercontent.com/57886661/264223579-17db8b93-d0c9-4a36-b088-39b944bb3d8d.png)
> 
> What OS and rust version (`rustc --version`) are you running the tests with?

This is my setup, I thought it was OS problem and I created a fresh instance with more RAM tho 
![image](https://github.com/sigp/lighthouse/assets/57886661/b0e06b19-df97-4b29-b261-f192dd850269)

michaelsproul commented:
@protocolwhisper Why is your username `autogpt`? Are you an LLM?

michaelsproul commented:
The memory-db crate isn't part of Lighthouse, so it's strange that you sent a screenshot with an error from it.

protocolwhisper commented:
> The memory-db crate isn't part of Lighthouse, so it's strange that you sent a screenshot with an error from it.

Not at all `autogpt`  it's just the name of the instance, that's rare I'm executing `make test-full` now I'm running it in a githubcodespace to see if it works cause I already tried a couple of times. I've already tried it in my Fedora main build and the autogpt that's an Ubuntu instance. Btw, here's the log where I encounter the error when running `make test-full` in the root directory (https://bafkreiauwmsqqun2twupd4vq44p5e6ynv7ahpju7dm7vinsdajjrkknsaa.ipfs.nftstorage.link/
![image](https://github.com/sigp/lighthouse/assets/57886661/dc552cef-e872-4b6e-a8ed-0c806cca5a12)

protocolwhisper commented:
Update : I got the same in the fresh githubcodespace :I , 
![image](https://github.com/sigp/lighthouse/assets/57886661/2c0ae67b-59d8-4141-938c-24f935b46d61)

michaelsproul commented:
@protocolwhisper Sorry about that, the username and the weird error had me wondering :sweat_smile: 

I think that panic must be coming from Anvil, because it _does_ depend on `memory-db`, here:

https://github.com/foundry-rs/foundry/blob/6c4c68a7031581bb8b8a10bb44db8dff4e04277f/Cargo.lock#L3992-L4001

I think it's probably a buggy version that was recently released, which hasn't broken our CI yet because we're pinned to an older version. We'll sort it out, and in the meantime you can make progress on your PR by just running the validator client tests:

```
cargo test --release -p validator_client
```

That should give sufficient coverage for your change, then you can push here to see if it passes CI.

jimmygchen commented:
Nice, that was great find @michaelsproul ! Our CI actually got stuck because we're using a nightly version of Anvil (it's fixed now). Thanks!

jimmygchen commented:
FYI we're now using Anvil version `nightly-ca67d15f4abd46394b324c50e21e66f306a1162d`, which can be installed using
`foundryup --version nightly-ca67d15f4abd46394b324c50e21e66f306a1162d`

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
(rebase) : Convert Rejection into signed Response

protocolwhisper commented:
Hello, sorry for the delay. I recently began an internship and haven't had much time. I've already made some changes to the code. It's passing the tests, except for one. `http_api::tests::keystores::import_wrong_number_of_passwords`
![image](https://github.com/sigp/lighthouse/assets/57886661/bfe928d4-962d-4130-adfe-e26a1924730e)
I get this by executing `cargo test --release -p validator_client` btw 
Let me know how I can proceed, cause I don't get how key stores test is failing and backtrace doesn't help that much :)

jimmygchen commented:
Hi @protocolwhisper, 

Thanks for getting back to this PR! and congrats on the new internship ☺️ 

The error you posted is actually quite useful. If you look at the failing line `keystores.rs:351`, it's indicates that the endpoint expected a `400` response status code, but it's getting a `500` status code instead:

https://github.com/sigp/lighthouse/blob/1373dcf076bd031e1c09119281c5f84b789974c2/validator_client/src/http_api/tests/keystores.rs#L351

This is because it's being set in your `convert_rejection` function here:

https://github.com/sigp/lighthouse/pull/4606/files#diff-d609f87c391f5d418021060636431414389b24a447043bfde82e1ad9759916c5R1251-R1252

Instead we should probably use the `warp_utils::reject::handle_rejection` function to convert the rejections into `Reply`s, and only return `INTERNAL_SERVER_ERROR` if it's an unhandled rejection, similar to how we're doing this in the `beacon_node` apis:

https://github.com/sigp/lighthouse/blob/8e95b69a1a9a449f352ec7a3904c549accb75111/beacon_node/http_api/src/task_spawner.rs#L45-L52

jimmygchen reviewed (changes requested):
I've added a comment above, hope it helps :)

protocolwhisper committed:
(rebase) : Improve rejection error handling

protocolwhisper committed:
(rebase): Improving rejection error handling

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper commented:
> I've added a comment above, hope it helps :)

Thank you! 😄 I was lost trying to understand that test. I've made some changes, and now it's passing `cargo test --release -p validator_client` 🔥

protocolwhisper committed:
(rebase): Fixing comments

protocolwhisper commented:
@jimmygchen Hello , I spotted that `execution-engine-integration-ubuntu (pull_request)` it's not passing any comments on this ?

jmcph4 commented:
> @jimmygchen Hello , I spotted that `execution-engine-integration-ubuntu (pull_request)` it's not passing any comments on this ?

Consult #4836.

protocolwhisper commented:
Since https://github.com/sigp/lighthouse/pull/4836 got merged, can I get a re-run for the PR?

protocolwhisper commented:
Since https://github.com/sigp/lighthouse/pull/4836 got merged, can I get a re-run for the tests? @jimmygchen @michaelsproul 😄

michaelsproul commented:
@protocolwhisper Merge unstable to your branch and we'll re-run

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

jimmygchen commented:
Feel free to ignore the test failure, it seems to be related to an issue we discovered earlier today, which is now being addressed in https://github.com/sigp/lighthouse/pull/4868. I'll re-run the fail jobs now.

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

jimmygchen commented on common/warp_utils/src/task.rs:
```diff
@@ -11,7 +11,6 @@ where
         .await
         .unwrap_or_else(|_| Err(warp::reject::reject()))
 }
-
```
Could you please add this empty line back?

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1108,7 +1109,7 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
+        .then(|request, signer, validator_store, task_executor, log| {
             blocking_signed_json_task(signer, move || {
                 remotekeys::delete(request, validator_store, task_executor, log)
             })
```
There's another occurrence of `and_then` here:
https://github.com/sigp/lighthouse/blob/f77a5a0a4e7bc565bb7079998eb81c2bdd85ccf4/validator_client/src/http_api/mod.rs#L1124

protocolwhisper replied:
Fixed

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1254,14 +1251,43 @@ where
+                Err(_) => Response::builder()
+                    .status(StatusCode::INTERNAL_SERVER_ERROR)
+                    .body(vec![])
+                    .expect("unhandled error in blocking task"),
```
I know this is an existing pattern here, but it would be nice to get rid of these `expect` usage even though they are unreachable. An unexpected error here will result in a `panic` which is what we try very hard to avoid in Lighthouse.

I'd suggest we refactor these `Response::builder` usage here to the same pattern as our beacon node code to something like this:

https://github.com/sigp/lighthouse/blob/8e95b69a1a9a449f352ec7a3904c549accb75111/beacon_node/http_api/src/task_spawner.rs#L47-L52

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1254,14 +1251,43 @@ where
-
             response
-        })
+        }
```
This is bit isn't really doing what the function is advertised to do; I think it probably falls under the responsibility of the `blocking_signed_json_task` function instead as it describes here:

https://github.com/sigp/lighthouse/blob/f77a5a0a4e7bc565bb7079998eb81c2bdd85ccf4/validator_client/src/http_api/mod.rs#L1277-L1280

Is it possible to move this block there to keep the `convert_rejection` only do what it's supposed to do?

> Convert a warp `Rejection` into a `Response`.

If it's too much of an issue, we could consider renaming this function to something like `convert_into_json_response`?

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1227,22 +1228,18 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
-        .map(|func_output| {
-            let mut response = match serde_json::to_vec(&func_output) {
+    match res {
+        Ok(func_output) => {
```
`func_output` no longer make sense in the context of this function (`func` is a param in the `blocking_signed_json_task`), we should probably rename this if we're going to keep this bit in this function.

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1254,14 +1251,43 @@ where
+            //Building the response with the Rejection
+            match resp {
+                Ok(reply) => {
+                    let response = reply.into_response();
+                    let (_parts, body) = response.into_parts();
+                    let body_bytes = hyper::body::to_bytes(body).await.unwrap();
+                    Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(body_bytes.to_vec())
+                        .expect("can't produce response from rejection")
```
I'm not sure if I understand this, why do we convert this into a `BAD_REQUEST` response and throw away the status code from `handle_rejection`?

It looks like we're converting the body from json -> bytes -> vec here, which feels a bit unnecessary.

jimmygchen commented on validator_client/src/http_api/mod.rs:
```diff
@@ -1254,14 +1251,43 @@ where
+                Ok(reply) => {
+                    let response = reply.into_response();
+                    let (_parts, body) = response.into_parts();
+                    let body_bytes = hyper::body::to_bytes(body).await.unwrap();
```
Avoid using `unwrap` as it could result in panic. If this conversion is needed and it fails then we could return an internal server error.

jimmygchen reviewed (changes requested):
Hi @protocolwhisper 
Thanks for the updates, I've added a few more comments. Let me know if you have any questions.

protocolwhisper committed:
(rebase) : empty lines

protocolwhisper committed:
Merge branch 'sigp:unstable' into unstable

protocolwhisper committed:
(rebase) : Changed output to Response<body>

protocolwhisper committed:
(fix): logic & error handling

protocolwhisper commented:
Hi @jimmygchen 
I just updated both functions, changed the output and updated the error pattern as the one from beacon  node

dapplion committed:
Run cargo fmt

dapplion committed:
Clean doc

protocolwhisper commented:
Heyy , can I get any comments about this?

michaelsproul commented:
Sorry @protocolwhisper but we've decided to delete the response signing altogether, so this change has been rolled into the PR that implements that: https://github.com/sigp/lighthouse/pull/5529

protocolwhisper commented:
> Sorry @protocolwhisper but we've decided to delete the response signing altogether, so this change has been rolled into the PR that implements that: #5529

Got it, I'll close this issue then

protocolwhisper commented:
Got it, I'll close this issue then

jimmygchen commented:
@protocolwhisper Sorry this one didn't work out, thanks for your help though!  🙏
